src/System.IO.Pipelines/PipeCompletion.cs

```diff
@@ -38,21 +83,49 @@ public bool IsCompletedOrThrow()
             return false;
         }
 
-        private void ThrowFailed()
+        public void InvokeCallbacks()
         {
-            throw _exception;
+            Debug.Assert(IsCompleted);
+
+            if (_callbacks == null)
```

Comment:
These are fire once callbacks right? Why not clear out the list here after invoking?

```
var callbacks = _callbacks;
_callbacks = null;
```